### PR TITLE
Validator: cache domain data calls

### PR DESCRIPTION
**Pull request by prestonvanloon**
Gated behind feature flag `--enable-domain-data-cache`

Also adds a new gRPC interceptor for logging backend and request duration.

### CODE REVIEW

**Comment by prestonvanloon** [comment]
Needs to be tested more carefully at runtime.

File: validator/client/validator.go [hunk 1]
```diff
@@ -330,3 +338,50 @@ func (v *validator) isAggregator(ctx context.Context, committee []uint64, slot u
 
 	return binary.LittleEndian.Uint64(b[:8])%modulo == 0, nil
 }
+
+// UpdateDomainDataCaches by making calls for all of the possible domain data. These can change when
+// the fork version changes which can happen once per epoch. Although changing for the fork version
+// is very rare, a validator should check these data every epoch to be sure the validator is
```
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply validator/client/validator.go]: Data is the plural form of datum, so it is correct how it is already.

File: validator/client/service.go [hunk 2]
```diff
@@ -132,6 +134,15 @@ func (v *ValidatorService) Start() {
 	}
 
 	v.conn = conn
+	cache, err := ristretto.NewCache(&ristretto.Config{
+		NumCounters: 1280, // number of keys to track.
```
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply validator/client/service.go]: I made them up! 

128 seems reasonable number of keys to track, docs recommend x10 counters.

File: validator/client/validator.go [hunk 3]
```diff
@@ -330,3 +338,50 @@ func (v *validator) isAggregator(ctx context.Context, committee []uint64, slot u
...
+	if !featureconfig.Get().EnableDomainDataCache {
+		return
+	}
+
+	for _, d := range [][]byte{
+		params.BeaconConfig().DomainRandao,
+		params.BeaconConfig().DomainBeaconAttester,
+		params.BeaconConfig().DomainBeaconProposer,
+	} {
+		_, _ = v.domainData(ctx, helpers.SlotToEpoch(slot), d)
```
[the remark this answers is not in the document]
Comment (prestonvanloon) [reply validator/client/validator.go]: It couldn't hurt, i'll add a log, thanks!

File: validator/client/validator.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (0xKiwi) [reply validator/client/validator.go]: English is weird

**Pull request merged**